src/PackagePrivate/Humanizer/InternationalizedHumanizer.php

```diff
@@ -91,22 +92,86 @@ private function humanizeSeason( Season $season ): string {
 		);
 	}
 
-	private function humanizeDate( ExtDate $date ): string {
-		$humanizedDate = $this->humanizeDateWithoutUncertainty( $date );
+	/**
+	 * @param string[] $parts
+	 */
+	private function humanSeparator( array $parts ) : string {
+		if ( !count( $parts ) ) {
+			return "";
+		}
 
-		if ( $date->getQualification()->isApproximate() && $date->getQualification()->uncertain() ) {
-			return $this->message( 'edtf-maybe-circa', $humanizedDate );
+		if ( count( $parts ) === 1 ) {
+			return current( $parts );
 		}
 
-		if ( $date->getQualification()->isApproximate() ) {
-			return $this->message(
-				'edtf-circa',
-				!$this->languageStrategy->monthUppercaseFirst() ? strtolower( $humanizedDate ) : $humanizedDate
-			);
+		$last = array_pop( $parts );
+		return implode( ', ', $parts ) . $this->message( 'edtf-and' ) . $last;
+	}
+
+	private function humanizedDateByMessage( string $humanizedDate, string $msgKey ) : string {
+		if ( $this->languageStrategy->monthUppercaseFirst()
+			|| strpos( $this->message( $msgKey ), '$' ) === 0 ) {
+			return $humanizedDate;
+		}
+		
+		return strtolower( $humanizedDate );
+	}
+
+	private function composeMessage( ExtDate $date, string $humanizedDate ) : string {
+		$qualification = $date->getQualification();
+
+		$parts = [
+			'uncertain' => $qualification->getUncertainParts(),
+			'approximate' => $qualification->getApproximateParts(),
+			'uncertain-and-approximate' => $qualification->getUncertainAndApproximateParts(),
+		];
+
+		// this data-structure, together with the loop below
+		// cannot be moved to the Qualification class since
+		// Qualification::UNDEFINED does NOT return the parts
+		// of the date that are NULL. So where this is not
+		// suitable to the Qualification class (i.e. it does
+		// not fit the purpose of that class) some of the logic
+		// in this method could be moved to the ExtDate class itself
+			
+		$undefinedParts = array_filter( [
+			$date->getDay() === NULL,
+			$date->getMonth() === NULL,
+			$date->getYear() === NULL
+		] );
+
+		// check if whole date is uncertain, approximate, or uncertain and approximate
```

Comment:
Comments that explain what part of a function does are a big hint that the function is doing too much, and that the part can be extracted. This tends to remove the need for the comment.